kore/src/Kore/Step/Simplification/OrPattern.hs

```diff
@@ -36,4 +47,43 @@ simplifyPredicatesWithSmt unsimplified = do
     simplifiedOrs <- traverse
         (BranchT.gather . Pattern.simplifyPredicate)
         (MultiOr.extractPatterns unsimplified)
-    SMT.Evaluator.filterMultiOr (MultiOr.make (concat simplifiedOrs))
+    filterMultiOrWithTermCeil (MultiOr.make (concat simplifiedOrs))
+
+filterMultiOrWithTermCeil
+    :: forall simplifier variable
+    .   ( FreshVariable variable
+        , MonadSimplify simplifier
+        , Ord variable
+        , Show variable
+        , SortedVariable variable
+        , Unparse variable
+        )
+    => OrPattern variable -> simplifier (OrPattern variable)
+filterMultiOrWithTermCeil unfiltered = do
+    wrapped <- wrapAndAddCeils unfiltered
+    filtered <- SMT.Evaluator.filterMultiOr wrapped
+    return (fmap unwrap filtered)
+  where
+    wrapAndAddCeils
```

Comment:
When we add the `\ceil(term)`, why doesn't the result become an `OrPredicate`?

Reply:
Well, when we compute `\ceil(term)`, the result is, indeed, a predicate (i.e. `ceils` below is an `OrPredicate`).

However, I need to remove the results which become bottom because of that `ceil`, so I'm transforming a `term + predicate` into a `(term + predicate) + (predicate and ceil)`, i.e. with `wrapAndAddCeil` I'm transforming a `Pattern` into a `Conditional (Pattern)`, and with `wrapAndAddCeils` I'm transforming an `OrPattern` into a `MultiOr (Conditional (Pattern))`.

I can then use the SMT to filter this `MultiOr`, then I can extract the term (which is the original `Pattern`) to get the filtered `OrPattern` that I need to return.

The other option would have been to return a `MultiOr (Pattern, Predicate)`, but I can't see an easy way to make this work with just `MultiOr Predicate`.

Reply:
Oh, I see: the outer `Conditional` contains the `\ceil`, and also the predicates of the inner `Pattern`, is that right?

Reply:
Yes.